src/Microsoft.DotNet.Build.Tasks.Feed/BlobFeedAction.cs

```diff
@@ -20,86 +22,87 @@ sealed class BlobFeedAction
                 }
 
-                using (var clientThrottle = new SemaphoreSlim(this.MaxClients, this.MaxClients))
-                {
-                    await Task.WhenAll(items.Select(item => PushItemToFeed(item, feed.RelativePath, clientThrottle, allowOverwrite, false)));
-                }
+                await PushItemsToFeedAsync(items, allowOverwrite);
             }
 
             return !Log.HasLoggedErrors;
         }
 
-        public async Task<bool> PushToFeedFlat(IEnumerable<string> items, bool allowOverwrite = false)
+        public async Task<bool> PushItemsToFeedAsync(IEnumerable<string> items, bool allowOverwrite)
         {
-            if (CancellationToken.IsCancellationRequested)
-            {
-                Log.LogError("Task PushToFeedFlat cancelled");
-                CancellationToken.ThrowIfCancellationRequested();
-            }
+            Log.LogMessage(MessageImportance.Low, $"START pushing items to feed");
 
-            using (var clientThrottle = new SemaphoreSlim(this.MaxClients, this.MaxClients))
-            {
-                await Task.WhenAll(items.Select(item => PushItemToFeed(item, feed.RelativePath, clientThrottle, allowOverwrite, true)));
-            }
-            return !Log.HasLoggedErrors;
-        }
-
-        public async Task<bool> PushItemToFeed(string item, string relativePath, SemaphoreSlim clientThrottle, bool allowOverwrite, bool isFlat)
-        {
             try
             {
-                string uploadPath = feed.CalculateBlobPath(item, relativePath);
-                string packageDirectory = feed.CalculateRelativeUploadPath(item, relativePath);
-
-                await UploadAsync(CancellationToken, item, uploadPath, clientThrottle, allowOverwrite);
-
-                if (!isFlat)
+                // In case the first Push attempt fails with an InvalidOperationException we Init the feed and retry the Push command once.
+                // Sleet internally retries 5 times on each package when the push operation fails so we don't need to retry ourselves.
+                for (int i = 0; i <= 1; i++)
                 {
-                    List<string> listAzureBlobs = await ListAzureBlobs.ListBlobs(Log, feed.AccountName, feed.AccountKey, feed.ContainerName, packageDirectory);
-                    if (!listAzureBlobs.Any(x => x.Contains(uploadPath)))
+                    try
                     {
-                        throw new Exception($"Uploaded package {uploadPath} is not present on feed. Cannot update index.json.");
+                        bool result = await PushAsync(items.ToList(), allowOverwrite);
+                        return result;
+                    }
+                    catch (InvalidOperationException ex) when (ex.Message.Contains("init"))
+                    {
+                        Log.LogWarning($"Sub-feed {source.FeedSubPath} has not been initialized. Initializing now...");
+                        bool result = await InitAsync();
```

Comment:
Doing this InitAsync in the catch block is probably also not a good idea. Does sleet not have a way to check to see if the feed is initialized?

Reply:
@jcagme please do this in a clean-up change.